src/txmempool.cpp

```diff
@@ -298,6 +300,27 @@ void CTxMemPoolEntry::SetDirty()
     nModFeesWithDescendants = GetModifiedFee();
 }
 
+size_t CTxMemPoolEntry::GetUnpaidActionCount() const
+{
+    if (tx->IsCoinBase()) {
+        return 0;
+    } else {
+        return std::max(
+            size_t {0},
+            std::max(GRACE_ACTIONS, tx->GetLogicalActionCount()) - (size_t)(nFee / MARGINAL_FEE));
+    }
+}
+
+// Return a fixed-point representation of the entry's weight ratio, where 1 is represented by WEIGHT_RATIO_SCALE.
+int128_t CTxMemPoolEntry::GetWeightRatio() const
+{
+    // ensure that the result will always be nonzero
+    static_assert(WEIGHT_RATIO_SCALE > MAX_MONEY);
+    return std::min(
+        int128_t {WEIGHT_RATIO_SCALE} * std::max(CAmount {1}, nFee) / tx->GetConventionalFee(),
```

Comment:
```suggestion
        (int128_t {WEIGHT_RATIO_SCALE} * std::max(CAmount {1}, nFee)) / tx->GetConventionalFee(),
```
nonblocking, I just hate having to think about operator precedence.